File: common/src/main/java/io/github/moulberry/moulconfig/xml/XMLContext.java
package io.github.moulberry.moulconfig.xml;

import io.github.moulberry.moulconfig.gui.GuiComponent;
import io.github.moulberry.moulconfig.internal.CollectionUtils;
import io.github.moulberry.moulconfig.internal.Warnings;
import io.github.moulberry.moulconfig.observer.GetSetter;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.var;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import javax.xml.XMLConstants;
import javax.xml.namespace.QName;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

@RequiredArgsConstructor
@Getter
public class XMLContext<T> {
    final XMLUniverse universe;
    final T object;

    public GuiComponent getChildFragment(Element element) {
        return CollectionUtils.getSingleOrThrow(getChildFragments(element, this));
    }

    public GuiComponent getChildFragment(Element element, Object rebind) {
        return CollectionUtils.getSingleOrThrow(getChildFragments(element, new XMLContext<>(universe, rebind)));
    }

    public List<GuiComponent> getChildFragments(Element element) {
        return getChildFragments(element, this);
    }

    public List<GuiComponent> getChildFragments(Element element, Object rebind) {
        return getChildFragments(element, new XMLContext<>(universe, rebind));
    }

    public List<GuiComponent> getChildFragments(Element element, XMLContext<?> context) {
        NodeList childNodes = element.getChildNodes();
        List<GuiComponent> list = new ArrayList<>();
        for (int i = 0; i < childNodes.getLength(); i++) {
            Node item = childNodes.item(i);
            if (item instanceof Element) {
                GuiComponent loadedFragment = universe.load(context, (Element) item);
                list.add(loadedFragment);
            }
        }
        return list;
    }

    public <E> E getPropertyFromAttribute(Element element, QName name, Class<E> type, E def) {
        GetSetter<E> prop = getPropertyFromAttribute(element, name, type);
        if (prop == null) {
            return def;
        }
        return prop.get();
    }

    private String getRawXMLValue(Element element, QName name) {
        if (!name.getNamespaceURI().equals(XMLConstants.NULL_NS_URI)) {
            Warnings.warn("Attributes should not have a namespace attached to them. This namespace will be ignored");
        }
        var attributeValue = element.getAttribute(name.getLocalPart());
        if (attributeValue.isEmpty()) return null;
        return attributeValue;
    }

    public <E> @Nullable GetSetter<E> getPropertyFromAttribute(@NotNull Element element, @NotNull QName name, @NotNull Class<E> type) {
        var attributeValue = getRawXMLValue(element, name);
        if (attributeValue == null) return null;
        if (attributeValue.startsWith("@")) {
            return getBoundProperty(attributeValue.substring(1), type);
        }
        var e = universe.mapXMLObject(attributeValue, type);
        return new GetSetter<E>() {
            @Override
            public E get() {
                return e;
            }

            @Override
            public void set(E newValue) {
                throw new UnsupportedOperationException();
            }
        };
    }


    public <E> Consumer<E> getMethodFromAttribute(Element element, QName name, Class<E> type) {
        String attribute = getRawXMLValue(element, name);
        if (attribute == null) return e -> {
        };
        if (!attribute.startsWith("@"))
            throw new RuntimeException("Object bound method without @ prefix " + attribute + " at " + name);
        return getBoundMethod(attribute.substring(1), type);
    }

    public Runnable getMethodFromAttribute(Element element, QName name) {
        String attribute = getRawXMLValue(element, name);
        if (attribute == null) return () -> {
        };
        if (!attribute.startsWith("@"))
            throw new RuntimeException("Object bound method without @ prefix " + attribute + " at " + name);
        return getBoundMethod(attribute.substring(1));
    }

    public <E> Consumer<E> getBoundMethod(String name, Class<E> argument) {
        return universe.getPropertyFinder(object.getClass()).getBoundFunction(name, object, argument);
    }

    public Runnable getBoundMethod(String name) {
        return universe.getPropertyFinder(object.getClass()).getBoundFunction(name, object);
    }

    public <E> GetSetter<E> getBoundProperty(String name, Class<E> type) {
        var propertyFinder = universe.getPropertyFinder(object.getClass());
        return propertyFinder.getBoundProperty(name, type, object);
    }

}

File: common/src/main/java/io/github/moulberry/moulconfig/xml/XMLContext.kt
package io.github.moulberry.moulconfig.xml

import io.github.moulberry.moulconfig.gui.GuiComponent
import io.github.moulberry.moulconfig.internal.CollectionUtils
import io.github.moulberry.moulconfig.internal.Warnings
import io.github.moulberry.moulconfig.observer.GetSetter
import lombok.Getter
import lombok.RequiredArgsConstructor
import org.w3c.dom.Element
import java.util.function.Consumer
import javax.xml.XMLConstants
import javax.xml.namespace.QName

@RequiredArgsConstructor
@Getter
class XMLContext<T : Any>(
    val universe: XMLUniverse,
    val boundObject: T,
) {
    fun getChildFragment(element: Element): GuiComponent {
        return CollectionUtils.getSingleOrThrow(getChildFragments(element, this))
    }

    fun getChildFragment(element: Element, rebind: Any): GuiComponent {
        return CollectionUtils.getSingleOrThrow(getChildFragments(element, XMLContext(universe, rebind)))
    }

    fun getChildFragments(element: Element): List<GuiComponent> {
        return getChildFragments(element, this)
    }

    fun getChildFragments(element: Element, rebind: Any): List<GuiComponent> {
        return getChildFragments(element, XMLContext(universe, rebind))
    }

    fun getChildFragments(element: Element, context: XMLContext<*>): List<GuiComponent> {
        val childNodes = element.childNodes
        val list: MutableList<GuiComponent> = ArrayList()
        for (i in 0 until childNodes.length) {
            val item = childNodes.item(i)
            if (item is Element) {
                val loadedFragment = universe.load(context, item)
                list.add(loadedFragment)
            }
        }
        return list
    }

    fun <E> getPropertyFromAttribute(element: Element, name: QName, type: Class<E>, def: E): E {
        val prop = getPropertyFromAttribute(element, name, type) ?: return def
        return prop.get()
    }

    private fun getRawXMLValue(element: Element, name: QName): String? {
        if (name.namespaceURI != XMLConstants.NULL_NS_URI) {
            Warnings.warn("Attributes should not have a namespace attached to them. This namespace will be ignored")
        }
        val attributeValue = element.getAttribute(name.localPart)
        return if (attributeValue.isEmpty()) null else attributeValue
    }

    fun <E> getPropertyFromAttribute(element: Element, name: QName, type: Class<E>): GetSetter<E>? {
        val attributeValue = getRawXMLValue(element, name) ?: return null
        if (attributeValue.startsWith("@")) {
            return getBoundProperty(attributeValue.substring(1), type)
        }
        val e = universe.mapXMLObject(attributeValue, type)
        return object : GetSetter<E> {
            override fun get(): E {
                return e
            }

            override fun set(newValue: E) {
                throw UnsupportedOperationException()
            }
        }
    }

    fun <E> getMethodFromAttribute(element: Element, name: QName, type: Class<E>): Consumer<E> {
        val attribute = getRawXMLValue(element, name) ?: return Consumer { }
        if (!attribute.startsWith("@")) throw RuntimeException("Object bound method without @ prefix $attribute at $name")
        return getBoundMethod(attribute.substring(1), type)
    }

    fun getMethodFromAttribute(element: Element, name: QName): Runnable {
        val attribute = getRawXMLValue(element, name) ?: return Runnable {}
        if (!attribute.startsWith("@")) throw RuntimeException("Object bound method without @ prefix $attribute at $name")
        return getBoundMethod(attribute.substring(1))
    }

    fun <E> getBoundMethod(name: String, argument: Class<E>): Consumer<E> {
        return universe.getPropertyFinder(boundObject.javaClass).getBoundFunction(name, boundObject, argument)
    }

    fun getBoundMethod(name: String): Runnable {
        return universe.getPropertyFinder(boundObject.javaClass).getBoundFunction(name, boundObject)
    }

    fun <E> getBoundProperty(name: String, type: Class<E>): GetSetter<E> {
        val propertyFinder = universe.getPropertyFinder(boundObject.javaClass)
        return propertyFinder.getBoundProperty(name, type, boundObject)
    }
}
